Add counterexamples for parser errors

keltono opened the pull request:
This is the culmination of the work I mostly did 2 summers ago and have been finishing up since when I could.
The code is (I would say) well commented, and is mostly contained within `src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/compiletime/auxiliary/counterexample`. 
The way that the code is organized and the scope of the project made this feel like a reasonable way to do things --- if need be the code could be messaged into a builder/checker/dumper, but this doesn't feel like the correct abstraction.  

After some time I decided to restrict the scope of the algorithm to only produce non-unified counterexamples, i.e examples that may not have the exact same tokens. Getting this always is (of course) undecidable in general, and the algorithm presented in the literature is massively complex slow, and very brittle to changes. I think it would be essentially unmaintainable without significant reworking of the core algorithm. 

Let me know if they're any questions or concerns with anything.
I've added Lucas and August as reviewers, let me know if I should do something else or if the review/merge process should be done in a particular way.
CC @ericvanwyk @krame505 @schwerdf .

keltono committed:
first pass at nonUnifying counterexamples

keltono committed:
add back legacy

keltono committed:
fix incorrect setup

keltono committed:
fix transition table

keltono committed:
various bug fixes

keltono committed:
refactor and get partly working

keltono committed:
various

keltono committed:
non-unifying examples work

keltono committed:
cleanup

keltono committed:
cleanup debugging comments

keltono committed:
more cleanup

keltono committed:
add pretty printing of derivations

keltono committed:
add colour (not working, add a colourStringBuilder class)

keltono committed:
fix coloured counterexamples

keltono committed:
add reverseProduction/Transition methods

keltono committed:
add prepend, reduce1, prevSymbol, etc

keltono committed:
add reduce 2

keltono committed:
unified example first pass

keltono committed:
unpushed changes

keltono committed:
fix bug in non-unifying example

keltono committed:
works on one examplegit add -A!

keltono committed:
bug fixes and cleanup

keltono committed:
begin cleanup, refactoring, and documentation

keltono committed:
lots of documentation and refactoring

keltono committed:
more comments

keltono committed:
refactor a bunch, finish documentation

keltono committed:
bug fixes

keltono committed:
fix bugs in unified example search

keltono committed:
fix off by one errors

keltono committed:
Remove unified counterexample logic and complete doc comments

keltono committed:
add dot support, rework arguments, make style consistant

keltono committed:
Merge remote-tracking branch 'melt/develop' into NonUnified

keltono committed:
improve parse counterexample explanation

schwerdf commented on copper-parent.iml:
Was this file included deliberately?

schwerdf commented on src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompiler.java:
```diff
@@ -78,6 +78,8 @@ private static String usageMessage(ParserCompilerParameters args)
 		rv += "\t-mda\tRun Copper's modular determinism analysis on the input.\n\t\t\tIf this switch is used, the input must comprise exactly\n\t\t\ttwo grammars: the host and an extension to test.\n";
 		rv += "\t-avoidRecompile\tRun Copper only if one spec-file has a later\n\t\t\tmodification time than the output file.\n\t\t\tIf this switch is used, an output file must also\n\t\t\tbe specified.\n";
 		rv += "\t-logfile [lout]\tPipe all log output to the file 'lout'\n\t\t\t(default standard error).\n";
+		rv += "\t-dot [dout]\tOutput Graphiz dot code representing any counterexamples to 'dout'\n\t\t\t(no dot output if not provided).\n";
```
The name of Graphviz is consistently misspelled across this PR. I would resolve this by removing the Graphviz references from the names altogether, since the DOT format is supported by more than one graph layout tool.

keltono replied:
Makes sense, fixed.

schwerdf commented on src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/compiletime/pipeline/SpecParserParameters.java:
What was the reason for removing this interface?

keltono replied:
In order to pass some of the parameters needed for the new options,  there were parts of the code that used this interface, and some that used `ParserSpecCompiler` in a way that made it impossible (or at least difficult) to pass the information down the pipeline. Collapsing it to one interface made this a lot easier and I think will make adding any flags in the future easier.

schwerdf commented on src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/compiletime/skins/cup/CupSkinParser.java:
```diff
@@ -858,8 +858,9 @@ public java.lang.Object runSemanticAction_124()
-	                if(ignore) ignoreTerminals.add(CopperElementReference.ref(newTerminalName,name.location));
+
+	                //TODO was there a reason this didn't have a current grammar attached?
+	                if(ignore) ignoreTerminals.add(CopperElementReference.ref(currentGrammar.getName(),newTerminalName,name.location));
```
The CUP skin only supports a single grammar, so putting a grammar in the reference was not strictly necessary. However, since this is a generated file, it should be kept in sync with its source, [CupSkinGrammar.xml](https://github.com/melt-umn/copper/blob/develop/doc/specs/CupSkinGrammar.xml).

keltono replied:
Reverted.

schwerdf reviewed (changes requested):
Looks good in general. Specific concerns are in code comments.

krame505 commented on src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/compiletime/auxiliary/counterexample/CounterexampleSearch.java:
```diff
@@ -0,0 +1,601 @@
+ * For the sake of complexity, this implements only non-unifying counterexamples.
+ * @author Kelton OBrien
+ */
+public class CounterexampleSearch {
```
I'm not going to try to review the logic here.  But nothing looks obviously wrong to me.

schwerdf replied:
There are several debugging lines in here that print to `System.out`, and should be commented out.

krame505 commented on src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java:
```diff
@@ -192,6 +202,8 @@ public ParserCompilerParameters()
 		dumpFile = null;
 		dumpOutputType = null;
 		avoidRecompile = false;
+		colorCounterexample = true;
```
Would it be easy to make the default be to check whether stdout supports ANSI colors?

schwerdf replied:
The only way I know of to check terminal capabilities like this is OS-specific (reading the `TERM` environment variable). What could be done is to default to "no color" when the output is to a log file.

Also, I see that the Javadoc in a few places refers to the ANSI codes as "ascii color codes". This should be corrected as it goes into Copper's public documentation.

keltono replied:
Fixed the color code names. Working on changing the default when the output is a log file now.

keltono replied:
Changed, I had to add another flag to enable color to accommodate this, which I felt was better than not having an override to this behavior, or having a flag that takes a boolean as argument and needing to turn a string in to a boolean.

schwerdf replied:
We really should rework this so that the "default" color setting is the same whether invoking Copper through the CLI or the API, but we do not need to hold up the merge on this account.

krame505 commented on src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java:
```diff
@@ -333,6 +344,26 @@ public boolean isWarnUselessNTs()
+		this.colorCounterexample = colorCounterexample;
+	}
+
+	public void setGraphizDotOutput(File graphizDotOutput)
```
We should probably add a flag in Silver to enable this - or make it happen with `--copperdump`.  

Also I agree with August's naming comment - should probably just be `dotOutput`, unless you are doing something that is non-standard dot and specific to GraphViz.

keltono replied:
Is there a particular preference on which of those should be done? I could add it to copperdump without too much issue, but I haven't done so yet.

krame505 replied:
How much time does it add for a large parser, e.g. ableC?  I'd say add it to `--copperdump` if it's no more than a few seconds, otherwise make it a seperate flag.

keltono replied:
It's pretty fast, so I'll just add it to copperdump. I did find another (more or less unrelated) bug (that I thought I squashed...) when testing that again, so I'll fix that as well when I add it. Not sure how long that will take, but hopefully I will be done by the end of the week.

keltono replied:
Alright, I looked at it and realized that I think it would require a non-trivial refactor to add to copperdump, so it would be best to add the flag to silver if it's not too much of a pain.

krame505 reviewed (commented):
Looks good, but deferring to August to approve.

keltono committed:
rename dot output, restore generated code

keltono commented:
Apologies for the slow work, I started the PR right before midterms began (my mistake!), but I (believe I) have addressed all of the issues except for not emitting color codes by default for log files, which I will finish by the end of the day. Let me know if there's anything else.

keltono committed:
change default color options

krame505 commented on src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompiler.java:
```diff
@@ -78,8 +78,9 @@ private static String usageMessage(ParserCompilerParameters args)
-		rv += "\t-dot [dout]\tOutput Graphiz dot code representing any counterexamples to 'dout'\n\t\t\t(no dot output if not provided).\n";
-		rv += "\t-nc\t\tDo not use ascii color escape sequences to color any counterexamples\n";
+		rv += "\t-dot [dout]\tOutput dot markup code representing any counterexamples to 'dout'\n\t\t\t(no dot output if not provided).\n";
+		rv += "\t-nc \t\tDo ot use ANSI color escape sequences to color any counterexamples\n\t\t\t(color is enabled by default for standard out, by default if using a logfile).\n";
```
Should be "disabled by default if using a logfile"?

schwerdf replied:
As the default log output is *stderr*, not stdout (and as the question of log colorization is not necessarily limited to the counterexamples) I would use the following text for the `-cc` and `-nc` options, which should make clear what the defaults are:
```
-cc Use ANSI color escape sequences in log output, even when logging to a file
-nc Suppress ANSI color in log output, even when logging to the terminal
```

Also, on commit 52ffd6ea, the color test seems to be the wrong way around -- I tried it out with a conflicting grammar, and the counterexamples were without color on the terminal unless I used `-cc` and with color in the logfile unless I used `-nc`.

keltono replied:
I like that wording better, changed. 
Yep, looks like i put a `==` where i needed a `!=`, that also should be fixed now.

krame505 reviewed (commented):
Looks good to me; I'll let @schwerdf approve.

keltono committed:
fix compiler color options

keltono committed:
fix flag typos

keltono commented:
I think everything is ready to go for merging!
